linkerd/access-log/src/layer.rs

```diff
@@ -0,0 +1,122 @@
+use crate::AccessLog;
+use futures::{ready, TryFuture};
+use pin_project::pin_project;
+use std::future::Future;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use std::time::SystemTime;
+use tokio::sync::mpsc;
+use tracing::warn;
+
+/// A layer that adds access logging
+#[derive(Clone)]
+pub struct AccessLogLayer {
+    sink: Option<mpsc::Sender<AccessLog>>,
+}
+
+#[derive(Clone)]
+pub struct AccessLogContext<Svc> {
+    inner: Svc,
+    sink: Option<mpsc::Sender<AccessLog>>,
+}
+
+#[pin_project]
+pub struct ResponseFuture<F> {
+    state: Option<(AccessLog, mpsc::Sender<AccessLog>)>,
+
+    #[pin]
+    inner: F,
+}
+
+impl<Svc> tower::layer::Layer<Svc> for AccessLogLayer {
+    type Service = AccessLogContext<Svc>;
+
+    fn layer(&self, inner: Svc) -> Self::Service {
+        Self::Service {
+            inner,
+            sink: self.sink.clone(),
+        }
+    }
+}
+
+impl<Svc, B1, B2> tower::Service<http::Request<B1>> for AccessLogContext<Svc>
+where
+    Svc: tower::Service<http::Request<B1>, Response = http::Response<B2>>,
+{
+    type Response = Svc::Response;
+    type Error = Svc::Error;
+    type Future = ResponseFuture<Svc::Future>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Svc::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, request: http::Request<B1>) -> Self::Future {
+        let sink = match &self.sink {
+            Some(sink) => sink,
+            None => {
+                return ResponseFuture {
+                    state: None,
+                    inner: self.inner.call(request),
+                }
+            }
+        };
+
+        let t0 = SystemTime::now();
+
+        let host = request.headers().get("Host").cloned();
+
+        let trace_id = request
+            .headers()
+            .get("X-Amzn-Trace-Id")
+            .or_else(|| request.headers().get("X-Request-ID"))
+            .cloned();
+
+        let user_agent = request.headers().get("User-Agent").cloned();
+
+        let log = AccessLog {
+            uri: request.uri().clone(),
+            method: request.method().clone(),
+            status: Default::default(),
+            host,
+            user_agent,
+            trace_id,
```

Comment:
bit of a shame that we have to clone so many strings here (the URI, `host` header, `user-agent` header, and trace id).

it seems like it might be more efficient if we just formatted all of the fields we know at this point into a single _new_ string, and then appended the remaining fields to it when the request completes. that way, we'd be making 2 allocations (one for the string, and one to carry the trace ID along because it occurs at the end of the line), rather than 5 or so (4 for all these values, and then another for the string that we format when the request completes).

just a thought!

Reply:
I _think_ that HeaderValue internally uses Bytes, which is internally reference counted, so these _should_ just be atomic increments... But I could be wrong on that front

Reply:
Oh, you're right; I didn't look closely enough at the struct definition and assumed they were being accessed as strings. The URI is also represented using `Bytes`, so this is fine.

Never mind!